flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/PartitionShuffleDescriptor.java

```diff
@@ -0,0 +1,139 @@
+
+	/** The type of the partition. */
+	@Nonnull
+	private final ResultPartitionType partitionType;
+
+	/** The number of subpartitions. */
+	private final int numberOfSubpartitions;
+
+	/** The maximum parallelism. */
+	private final int maxParallelism;
+
+	/** Connection index to identify this partition of intermediate result. */
+	private final int connectionIndex;
+
+	public PartitionShuffleDescriptor(
+		@Nonnull IntermediateDataSetID resultId,
+		@Nonnull IntermediateResultPartitionID partitionId,
+		@Nonnull ResultPartitionType partitionType,
+		int numberOfSubpartitions,
+		int maxParallelism,
+		int connectionIndex) {
+
+		this.resultId = resultId;
+		this.partitionId = partitionId;
+		this.partitionType = partitionType;
+		this.connectionIndex = connectionIndex;
+
+		KeyGroupRangeAssignment.checkParallelismPreconditions(maxParallelism);
+		checkArgument(numberOfSubpartitions >= 1);
+		this.numberOfSubpartitions = numberOfSubpartitions;
+		this.maxParallelism = maxParallelism;
+	}
+
+	@Nonnull
+	public IntermediateDataSetID getResultId() {
+		return resultId;
+	}
+
+	@Nonnull
+	public IntermediateResultPartitionID getPartitionId() {
+		return partitionId;
+	}
+
+	@Nonnull
+	public ResultPartitionType getPartitionType() {
+		return partitionType;
+	}
+
+	public int getNumberOfSubpartitions() {
+		return numberOfSubpartitions;
+	}
+
+	public int getMaxParallelism() {
+		return maxParallelism;
+	}
+
+	public int getConnectionIndex() {
+		return connectionIndex;
+	}
+
+	@Override
+	public String toString() {
+		return String.format("PartitionShuffleDescriptor [result id: %s, "
+				+ "partition id: %s, partition type: %s, connection index: %d]",
+			resultId, partitionId, partitionType, connectionIndex);
+	}
+
+	// ------------------------------------------------------------------------
+
+	public static PartitionShuffleDescriptor from(IntermediateResultPartition partition, int maxParallelism) {
+
+		final IntermediateDataSetID resultId = partition.getIntermediateResult().getId();
+		final IntermediateResultPartitionID partitionId = partition.getPartitionId();
+		final ResultPartitionType partitionType = partition.getIntermediateResult().getResultType();
+
+		// The produced data is partitioned among a number of subpartitions.
+		//
+		// If no consumers are known at this point, we use a single subpartition, otherwise we have
+		// one for each consuming sub task.
+		int numberOfSubpartitions = 1;
+
+		if (!partition.getConsumers().isEmpty() && !partition.getConsumers().get(0).isEmpty()) {
+
+			if (partition.getConsumers().size() > 1) {
+				throw new IllegalStateException("Currently, only a single consumer group per partition is supported.");
+			}
+
+			numberOfSubpartitions = partition.getConsumers().get(0).size();
+		}
+
+		return new PartitionShuffleDescriptor(
+			resultId, partitionId, partitionType, numberOfSubpartitions, maxParallelism,
+			partition.getIntermediateResult().getConnectionIndex());
```

Comment:
I am not sure whether it is worth defining a private `connectionIndex` var in front for `partition.getIntermediateResult().getConnectionIndex()` to keep the same way as other vars.